### ISSUE TITLE
Further restricting what coercions are allowed on places of type `!`

### ISSUE THREAD

**Issue by compiler-errors** [rust-lang/rust#131297]
In #129392, I fixed a couple soundness holes having to do with the distinction between values and places and the `!` type, which has special divergence and coercion behavior in HIR typeck.

To do so, I implemented a heuristic called `expr_guaranteed_to_constitute_read_for_never` (and same for patterns, `pat_guaranteed...`), which determines whether an expression/pattern is *guaranteed* to perform a read of a value of type `!`, which triggers the special divergence and coercion behavior. Read the description of the PR to see what the behavior is after the PR.

There's still some open questions about what expressions and patterns should be considered reads in the HIR (such as reading structs with only field subpatterns that don't read, like `Struct { .. }`), but I'd like to separate that since the PR #129392 is strictly an improvement over the existing behavior.

**Comment by compiler-errors** [comment]
@lcnr also brought up a good point that `ref`/`ref mut` bindings probably also don't constitute reads. We don't allow coercions on `let` if `ref`/`ref mut` right now anyways, so it's less of an issue.

**Comment by RalfJung** [comment]
Cc @rust-lang/opsem

**Comment by digama0** [comment]
The question about struct and 1-variant enum matches also leads to the following formulation of the question:

Is the following UB?
```
#[repr(u8)]
enum Foo {
    Foo,
}
fn main() {
    let x: Foo = unsafe { std::mem::uninitialized() };
    let Foo::Foo = x;
}
```
I think the answer is no, it is defined behavior to do a discriminant read on a 1-variant enum because nothing needs to be read and there is no data to read in the first place (`Foo::Foo` is all padding, so the fact that it is uninit makes no difference).

If we agree on this, then I think we should say that struct patterns, and 1-variant enum patterns, should not be considered reads in this HIR analysis.

**Comment by compiler-errors** [comment]
> If we agree on this, then I think we should say that struct patterns, and 1-variant enum patterns, should not be considered reads in this HIR analysis.

Yep, that's approximately what I was thinking.

**Comment by jwong101** [comment]
> I think the answer is no, it is defined behavior to do a discriminant read on a 1-variant enum because nothing needs to be read and there is no data to read in the first place.

I could see that argument for `repr(Rust)` 1-variant enums. But the `Foo` enum has an explicit discriminant representation of type `u8` that can only have 0 as it's value(the default discriminant value). Codegen currently exploits that by slapping `range` metadata on parameters+returns of type `Foo`.

**Comment by RalfJung** [comment]
Yeah I think that code is definitely UB on the `let x: Foo = unsafe` line. It doesn't have a padding byte, it has a 1-byte discriminant byte, and that must be initialized.

**Comment by WaffleLapkin** [comment]
Is there any desire to change the current rustc behavior?

The current behavior is
- The only patterns that do not constitute a read for `!` are wildcards (`_`) and or patterns where any of the patterns don't constitute reads
- The only expressions which don't constitute reads are
  - `...` in `addr_of!(...)`
  - Match scrutinee where all patterns in all arms don't constitute reads for never
  - Let initializer expression where the pattern doesn't constitute reads for never
  - `...` in `(...).field` and `dst` in `dst = src`
    - I don't think either of these matters in practice

I don't love this check, but I don't really see how to make it better, without being able to properly distinguish places vs values in HIR.

There are only 2 weird consequences of this check that I can see:
1. Patterns that assume the type, but don't read anything (`Struct { .. }`, `(_, _)`, etc) are considered to be a read and thus allow coercion (this is detected as UB by miri)
   - I think this is fine, by assuming the type, user requires the coercion to happen
2. Or patterns with only some patterns being `_` does not consitute a read (`Struct { .. } | _`)
   - This makes more code error than strictly necessary, which seems fine

So I'm not sure if there is anything actionable here.

**Comment by RalfJung** [comment]
> Or patterns with only some patterns being _ does not consitute a read (Struct { .. } | _) 

This seems a bit odd, but it may be better than having this be order-dependent.

> This makes more code error than strictly necessary, which seems fine

How does *not* considering this a read make more code error? It means there's less UB because there's fewer reads, right?

**Comment by WaffleLapkin** [comment]
> > This makes more code error than strictly necessary, which seems fine
> 
> How does not considering this a read make more code error? It means there's less UB because there's fewer reads, right?

The way we avoid UB is that we don't insert `NeverToAny` coercions (on things that look like places and where we think it doesn't constitute a read), which means that code can fail to type check. For example ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=f356a748616f267b111f0ff4ead1f26c)):

```rust
let ptr = std::ptr::null::<!>();
let (Struct { .. } | _) = *ptr;
```
```
error[E0308]: mismatched types
 --> src/main.rs:7:10
  |
7 |     let (Struct { .. } | _) = *ptr;
  |          ^^^^^^^^^^^^^        ---- this expression has type `!`
  |          |
  |          expected `!`, found `Struct`
  |
  = note: expected type `!`
           found struct `Struct`
```

**Comment by RalfJung** [comment]
Oh I see. So considering more things a read is a breaking change as it makes more code UB, and considering fewer things a read is a breaking change as it means we insert fewer coercions? Gnarly. I wonder if there's a way to have some wiggle room in the middle...

**Comment by lcnr** [comment]
> Oh I see. So considering more things a read is a breaking change as it makes more code UB, and considering fewer things a read is a breaking change as it means we insert fewer coercions? Gnarly. I wonder if there's a way to have some wiggle room in the middle...

We can in theory FCW on missing `NeverToAny` coercions by doing
- when encounter a coercion we want to remove, we emit the coercion, but track the fact that we did a FCW-able `NeverToAny` coercion at this point
- add coercions from infer vars to other things (if these infer vars are from a FCW-able `NeverToAny` coercion and at this point, also set the original infer var to `!`)
- emit an FCW if one of the tracked `NeverToAny` coercions resulted in a type that is not just `!`
